### PR TITLE
Add decimal annotations

### PR DESCRIPTION
**Pull request by ahmedabdou14**
## Change Summary

Create some decimal.Decimal annotated types to replace deprecated `condecimal` methods.

## Related issue number
Closes [#8246 ](https://github.com/pydantic/pydantic/issues/8246)

## Checklist

* [x] The pull request title is a good summary of the changes - it will be used in the changelog
* [x] Unit tests for the changes exist
* [x] Tests pass on CI
* [x] Documentation reflects the changes where applicable
* [x] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

### CODE REVIEW

**Comment by ahmedabdou14** [comment]
This PR might require a new PR in `pydantic-core` to support `ConfigDict.ser_json_inf_nan='null'`

In this PR `ConfigDict.ser_json_inf_nan='null'`, will not result in a `null` value for `Deciml('Infinity')` after dumping the `model_dump_json`

**Review by samuelcolvin — COMMENTED** [review]
Humm, personally I don' think this is worth the churn and increased "magic" to users.

Personally, I'd rather people write out 

```py
amount: Annotated[Decimal, annotated_types.Gt(0)]
```
manually, they can always create aliases themselves.

If we did go ahead with any of this, we'd need to fix the markdown, currently it's more like rst.

**Comment by ahmedabdou14** [comment]
You got a point aswell tbh, i find it a useful feature to be consistent with existing annotations for `float`. I also noticed many devs don't know about python builtin `annotated_types`,  they mostly use Pydantic ready types.

Regarding the docs, can you elaborate on the md/rst fix needed? thank you

**Comment by Viicos** [comment]
> I also noticed many devs don't know about python builtin `annotated_types`

`annotated_types` is a library [on its own](https://github.com/annotated-types/annotated-types/), but I agree I don't think people know much about it :thinking:

File: pydantic/types.py [hunk 1]
```diff
@@ -1077,6 +1083,172 @@ class ConstrainedExample(BaseModel):
...
+    print(e)
+    '''
+    1 validation error for StrictDecimalModel
+    strict_decimal
+      Input should be an instance of Decimal [type=is_instance_of, input_value='1.0', input_type=str]
+    '''
+```
+"""
+InfiniteDecimal = Annotated[Decimal, AllowInfNan(True)]
+"""A decimal that can be infinite or NaN(``-Infinity``, ``Infinity``, or ``nan``).
```
Comment (Viicos) [inline pydantic/types.py:1236]:
```suggestion
"""A decimal that can be infinite or NaN(`-Infinity`, `Infinity`, or `nan`).
```
> Regarding the docs, can you elaborate on the md/rst fix needed? thank you

I think it was regarding double quotes

**Comment by ahmedabdou14** [comment]
> > I also noticed many devs don't know about python builtin `annotated_types`
> 
> `annotated_types` is a library [on its own](https://github.com/annotated-types/annotated-types/), but I agree I don't think people know much about it 🤔

Oh I didn't notice it was a library

**Comment by ahmedabdou14** [comment]
Sorry closed by mistake :)

File: pydantic/types.py (hunk 1 above)
Comment (ahmedabdou14) [reply pydantic/types.py]: Done

**Comment by sydney-runkle** [comment]
Closing now.

@ahmedabdou14, thanks for your work. Looks like you understand parts of the codebase quite well. I've marked lots of issues with `good first issue` and `help wanted` if you're interested in diving into other projects :).

**Pull request closed without merging**